lib/Service/ImageService.php

```diff
@@ -0,0 +1,676 @@
+	 * @param array $params
+	 * @return array
+	 */
+	private function simpleDownload(string $url, $resource, array $params = []): array {
+		$client = $this->clientService->newClient();
+		try {
+			$options = [
+				// does not work with sink if SSE is enabled
+				// 'sink' => $resource,
+				// rather use stream and write to the file ourselves
+				'stream' => true,
+				'timeout' => 0,
+				'headers' => [
+					'User-Agent' => 'Nextcloud Text',
+				],
+			];
+
+			if (count($params) > 0) {
+				$paramsContent = http_build_query($params);
+				$url .= '?' . $paramsContent;
+			}
+
+			$response = $client->get($url, $options);
+			$body = $response->getBody();
+			while (!feof($body)) {
+				// write ~5 MB chunks
+				$chunk = fread($body, 5000000);
+				fwrite($resource, $chunk);
+			}
+
+			return ['Content-Type' => $response->getHeader('Content-Type')];
+		} catch (ServerException | ClientException $e) {
+			//$response = $e->getResponse();
+			//if ($response->getStatusCode() === 401) {
+			$this->logger->warning('Impossible to download image', ['exception' => $e]);
+			return ['error' => 'Impossible to download image'];
+		} catch (ConnectException $e) {
+			$this->logger->error('Connection error', ['exception' => $e]);
+			return ['error' => 'Connection error'];
+		} catch (Throwable | Exception $e) {
+			$this->logger->error('Unknown download error', ['exception' => $e]);
+			return ['error' => 'Unknown download error'];
+		}
+	}
+
+	/**
+	 * Actually delete attachment files which are not pointed in the markdown content
+	 *
+	 * @param int $fileId
+	 * @return int The number of deleted files
+	 * @throws NotFoundException
+	 * @throws \OCP\Files\InvalidPathException
+	 * @throws \OCP\Files\NotPermittedException
+	 * @throws \OCP\Lock\LockedException
+	 */
+	public function cleanupAttachments(int $fileId): int {
+		$textFile = $this->rootFolder->getById($fileId);
+		if (count($textFile) > 0 && $textFile[0] instanceof File) {
+			$textFile = $textFile[0];
+			if ($textFile->getMimeType() === 'text/markdown') {
+				// get IDs of the files inside the attachment dir
+				$attachmentDir = $this->getOrCreateAttachmentDirectoryForFile($textFile);
+				$attachmentsByName = [];
+				foreach ($attachmentDir->getDirectoryListing() as $attNode) {
+					$attachmentsByName[$attNode->getName()] = $attNode;
+				}
+
+				$contentAttachmentNames = $this->getAttachmentNamesFromContent($textFile->getContent());
+
+				$toDelete = array_diff(array_keys($attachmentsByName), $contentAttachmentNames);
+				foreach ($toDelete as $name) {
+					$attachmentsByName[$name]->delete();
+				}
+				return count($toDelete);
+			}
+		}
+		return 0;
+	}
+
+
+	/**
+	 * Get attachment file names listed in the markdown file content
+	 *
+	 * @param string $content
+	 * @return array
+	 */
+	private function getAttachmentNamesFromContent(string $content): array {
+		$matches = [];
+		preg_match_all(
+			'/\!\[[^\[\]]+\]\(text:\/\/image\?[^)]*imageFileName=([^)&]+)\)/',
```

Comment:
The regex doesn't match image links with square brackets in the description. I had a quick look at [the relevant part of the CommonMark spec](https://spec.commonmark.org/0.30/#links) and indeed, square brackets are allowed (under limited circumstances).

php-markdown uses the following regex syntax for the link/image name/description part between squares:

```regex
(?>[^\[\]]+|\[(?>[^\[\]]+|\[(?>[^\[\]]+|\[(?>[^\[\]]+|\[(?>[^\[\]]+|\[(?>[^\[\]]+|\[\])*\])*\])*\])*\])*\])*
```

See [`$this->nested_brackets_re`](https://github.com/michelf/php-markdown/blob/f0911e737ddaad081803424203407a0f01189eca/Michelf/Markdown.php#L106) and [`doImages()`](https://github.com/michelf/php-markdown/blob/f0911e737ddaad081803424203407a0f01189eca/Michelf/Markdown.php#L693)

Reply:
This is tricky. The inserted `[photo.name.jpg](text://image?imageFileName=photo.name.jpg)` has to be valid from the renderer point of view. I'll first fix that and then adjust `getAttachmentNamesFromContent()`. Thanks for mentioning this as image insertion fails in many cases.

Reply:
The expression from php-markdown is now used in `getAttachmentNamesFromContent()`.
As a simple fix for the nested brackets, as we generate the link ourselves and the user is not supposed to manually edit it afterwards, brackets in image file names are removed in the link text.
There is no problem in the link itself as it is "url encoded".
This is done in c351e66607d23b1672603752b2bdd1dba83b59a6

Reply:
Perfect, I think that's fine for now :ok_hand: